lang/src/accounts/interface_account.rs

```diff
@@ -0,0 +1,331 @@
+/// those programs are not annotated with `#[account]` so you have to
+/// - create a wrapper type around the structs you want to wrap with InterfaceAccount
+/// - implement the functions required by InterfaceAccount yourself
+/// instead of using `#[account]`. You only have to implement a fraction of the
+/// functions `#[account]` generates. See the example below for the code you have
+/// to write.
+///
+/// The mint wrapper type that Anchor provides out of the box for the token program ([source](https://github.com/coral-xyz/anchor/blob/master/spl/src/token.rs))
+/// ```ignore
+/// #[derive(Clone)]
+/// pub struct Mint(spl_token::state::Mint);
+///
+/// // This is necessary so we can use "anchor_spl::token::Mint::LEN"
+/// // because rust does not resolve "anchor_spl::token::Mint::LEN" to
+/// // "spl_token::state::Mint::LEN" automatically
+/// impl Mint {
+///     pub const LEN: usize = spl_token::state::Mint::LEN;
+/// }
+///
+/// // You don't have to implement the "try_deserialize" function
+/// // from this trait. It delegates to
+/// // "try_deserialize_unchecked" by default which is what we want here
+/// // because non-anchor accounts don't have a discriminator to check
+/// impl anchor_lang::AccountDeserialize for Mint {
+///     fn try_deserialize_unchecked(buf: &mut &[u8]) -> Result<Self> {
+///         spl_token::state::Mint::unpack(buf).map(Mint)
+///     }
+/// }
+/// // AccountSerialize defaults to a no-op which is what we want here
+/// // because it's a foreign program, so our program does not
+/// // have permission to write to the foreign program's accounts anyway
+/// impl anchor_lang::AccountSerialize for Mint {}
+///
+/// impl anchor_lang::Owner for Mint {
+///     fn owner() -> Pubkey {
+///         // pub use spl_token::ID is used at the top of the file
+///         ID
+///     }
+/// }
+///
+/// // Implement the "std::ops::Deref" trait for better user experience
+/// impl Deref for Mint {
+///     type Target = spl_token::state::Mint;
+///
+///     fn deref(&self) -> &Self::Target {
+///         &self.0
+///     }
+/// }
+/// ```
+///
+/// ## Out of the box wrapper types
+///
+/// ### SPL Types
+///
+/// Anchor provides wrapper types to access accounts owned by the token programs. Use
+/// ```ignore
+/// use anchor_spl::token_interface::TokenAccount;
+///
+/// #[derive(Accounts)]
+/// pub struct Example {
+///     pub my_acc: InterfaceAccount<'info, TokenAccount>
+/// }
+/// ```
+/// to access token accounts and
+/// ```ignore
+/// use anchor_spl::token_interface::Mint;
+///
+/// #[derive(Accounts)]
+/// pub struct Example {
+///     pub my_acc: InterfaceAccount<'info, Mint>
+/// }
+/// ```
+/// to access mint accounts.
+#[derive(Clone)]
+pub struct InterfaceAccount<'info, T: AccountSerialize + AccountDeserialize + Clone> {
+    account: Account<'info, T>,
+    // The owner here is used to make sure that changes aren't incorrectly propagated
+    // to an account with a modified owner
+    owner: Pubkey,
```

Comment:
using `owner` as a the variable name here is probably ok although we do already have the account info `owner` and the `owner` trait on the Account type.

Reply:
Yeah, I wasn't sure of a better term, so I went with the simplest name. Maybe `loaded_owner` or `passed_owner` or `actual_owner` would be better names? I'm leaning toward just `owner`, but I can be easily swayed